torchtext/_extension.py

```diff
@@ -1,16 +1,66 @@
-def _init_extension():
-    import importlib
-    import os
+import os
+from pathlib import Path
+
+import torch
+from torchtext._internal import module_utils as _mod_utils  # noqa: F401
+
+_LIB_DIR = Path(__file__).parent / "lib"
+
+
+def _get_lib_path(lib: str):
+    suffix = "pyd" if os.name == "nt" else "so"
+    path = _LIB_DIR / f"{lib}.{suffix}"
+    return path
+
+
+def _load_lib(lib: str) -> bool:
+    """Load extension module
 
-    import torch
+    Note:
+        In case `torchtext` is deployed with `pex` format, the library file
+        is not in a standard location.
+        In this case, we expect that `libtorchtext` is available somewhere
+        in the search path of dynamic loading mechanism, so that importing
+        `_torchtext` will have library loader find and load `libtorchtext`.
+        This is the reason why the function should not raising an error when the library
+        file is not found.
 
-    # load the custom_op_library and register the custom ops
-    lib_dir = os.path.dirname(__file__)
-    loader_details = (importlib.machinery.ExtensionFileLoader, importlib.machinery.EXTENSION_SUFFIXES)
+    Returns:
+        bool:
+            True if the library file is found AND the library loaded without failure.
+            False if the library file is not found (like in the case where torchtext
+            is deployed with pex format, thus the shared library file is
+            in a non-standard location.).
+            If the library file is found but there is an issue loading the library,
+            (such as missing dependency) then this function raises the exception as-is.
 
-    extfinder = importlib.machinery.FileFinder(lib_dir, loader_details)
-    ext_specs = extfinder.find_spec("_torchtext")
-    if ext_specs is None:
+    Raises:
+        Exception:
+            If the library file is found, but there is an issue loading the library file,
+            (when underlying `ctype.DLL` throws an exception), this function will pass
+            the exception as-is, instead of catching it and returning bool.
+            The expected case is `OSError` thrown by `ctype.DLL` when a dynamic dependency
+            is not found.
+            This behavior was chosen because the expected failure case is not recoverable.
+            If a dependency is missing, then users have to install it.
+    """
+    path = _get_lib_path(lib)
+    if not path.exists():
+        return False
+    torch.ops.load_library(path)
+    torch.classes.load_library(path)
```

Comment:
FYI: The implementation beneath these two different `load_library` are common, so only one of them are required.

Reply:
Hmm so I noticed audio actually makes a call to both. Any specific reason why it's done that way in torchaudio?

https://github.com/pytorch/audio/blob/main/torchaudio/_extension.py#L51-L52

Reply:
Yeah, I learned this fact myself only after I wrote this code. I have not updated audio yet.

Reply:
Okay let me go with `torch.ops.load_library` since that seems to be the approach vision also uses https://github.com/pytorch/vision/blob/9f48c6274f66f71dd0cabe6a48bde84c8f239abe/torchvision/extension.py#L20